### PR TITLE
Allow reading/updating/deleting a PreparedQuery by exact name

### PR DESCRIPTION
**Pull request by jamestran201**
Related to: #10531

## Background context

When a PreparedQuery is created and managed by a tool such as Ansible, its ID is not known ahead of time. This can make it difficult for the user to manage this query [in their scripts](https://github.com/hashicorp/consul/issues/10531#issuecomment-882480893). For this use case, the user currently manages these queries [manually](https://github.com/hashicorp/consul/issues/10531#issuecomment-882550258).

The issue above proposes adding endpoints to read/update/delete a query using its exact name to help the user with this use case, see [this comment](https://github.com/hashicorp/consul/issues/10531#issuecomment-895598803) for more details.

## Changes in this PR

This PR adds the ability to get a query using its exact name. The implementation is adapted from the code that retrieves a query by its ID.

The current changes in this PR:
- Add a REST endpoint to retrieve the query using its exact name. The endpoint path is `/v1/query/exact-name/{query-name}`.
- Add a RPC endpoint to retrieve the query using its exact name. This RPC endpoint is called `PreparedQuery.GetByExactName`.
- Add a function to retrieve the query from the state store using its name.
- Add a function to the `api` package to call the new endpoint
- Add API doc for `GET` endpoint

## To do
- Implement endpoint to update a query given its name
- Implement endpoint to delete a query given its name

## Questions

- ~~The implementation for reading a query given its name has a lot of duplication with reading the query by its ID. Should the common code be extracted?~~ Answered by https://github.com/hashicorp/consul/pull/12053#discussion_r795790995
- ~~Should the code for updating and deleting a query be implemented in this PR or separate ones?~~ Answered by https://github.com/hashicorp/consul/pull/12053#discussion_r795790995

### CODE REVIEW

**Comment by kisunji** [comment]
As for your questions, 
1: We're fine with duplicate code for now; we prefer not to prematurely refactor code until necessary (to avoid coupling code that may diverge later - we might see some of that with ACL auth here)
2: This PR is fine; update/delete can come after we've reviewed the ACL implications in this PR.

Thank you for your patience!

**Comment by FFMMM** [comment]
Hey @tmnhat2001 thanks for your patience here.

Our inclination is that we probably want to have ACL permissions checking for prepared query read when getting a prepared query by name but we want to do more research into this.

File: agent/consul/prepared_query_endpoint.go [hunk 1]
```diff
@@ -267,6 +267,43 @@ func (p *PreparedQuery) Get(args *structs.PreparedQuerySpecificRequest,
 		})
 }
 
+// GetByExactName returns a single prepared query by Name.
+func (p *PreparedQuery) GetByExactName(args *structs.PreparedQuerySpecificRequest,
+	reply *structs.IndexedPreparedQueries) error {
+	if done, err := p.srv.ForwardRPC("PreparedQuery.GetByExactName", args, reply); done {
+		return err
+	}
+
```
Comment (markan) [inline agent/consul/prepared_query_endpoint.go:276]: So I think we want to check for PreparedQueryRead permission here.

The setup logic for the auth could look similar to this:
https://github.com/hashicorp/consul/blob/8a6e75ac815a5bdfd9fdbd3528c985d5fdde9c1b/agent/consul/acl_endpoint.go#L1411

**Comment by jamestran201** [comment]
@markan @kisunji Sorry for the late reply 🙇‍♂️ . I recently changed jobs so haven't been able to work on this. Thank you for you guidance with ACL! I will start implementing it.

File: agent/consul/prepared_query_endpoint.go (hunk 1 above)
Comment (jamestran201) [reply agent/consul/prepared_query_endpoint.go:276]: Thank you for the example 👍 . I attempted to check the PreparedQueryRead permission in https://github.com/hashicorp/consul/pull/12053/commits/a08da226c1c4bb835821e5e3b49c812e9eb75d34. Please review it again when you have time.

**Pull request closed without merging**